### PR TITLE
 AP_Common: Make hexadecimal character number conversion method common

### PR DESCRIPTION
**Pull request by muramura**
There are multiple hex character number conversion methods that are the same.
Therefore we make it a common method.
Remove the local hexadecimal character number conversion method.
Change to common hexadecimal character number conversion method.

### CODE REVIEW

File: libraries/AP_Common/AP_Common.cpp [hunk 1]
```diff
@@ -36,3 +36,19 @@ bool is_bounded_int32(int32_t value, int32_t lower_bound, int32_t upper_bound)
...
 
 }
+
+/**
+ * return the numeric value of an ascii hex character
+ *
+ * @param [in] a hexa character
+ * @return numeric value
+ */
+ int16_t char_to_hex(int8_t a)
```
Comment (peterbarker) [inline libraries/AP_Common/AP_Common.cpp]: This is a very strange return type.  I wonder if this was originally intended to return `-1` on error?

Regardless, this should be `uint8_t`, right?

Comment (peterbarker) [reply libraries/AP_Common/AP_Common.cpp]: I know this implementation isn't yours, but it could be made more efficient.  10 times in 16 the LAST clause here will be taken - meaning we've done more comparisons than needed.  Something to fix later, perhaps.

File: libraries/AP_GPS/AP_GPS_NMEA.cpp [hunk 2]
```diff
@@ -215,7 +202,7 @@ bool AP_GPS_NMEA::_have_new_message()
         now - _last_GGA_ms > 150) {
         return false;
     }
-    if (_last_VTG_ms != 0 &&
```
Comment (peterbarker) [inline libraries/AP_GPS/AP_GPS_NMEA.cpp]: Unrelated whitespace change

File: libraries/AP_Common/AP_Common.cpp (hunk 1 above)
Comment (peterbarker) [reply libraries/AP_Common/AP_Common.cpp]: ... also, not a particularly apt name.

How about `hexchar_to_int`, or possibly just `unhex` (`hex` being what you use to go the other way in a lot of languages)

File: libraries/AP_GPS/AP_GPS_NMEA.cpp (hunk 2 above)
Comment (muramura) [reply libraries/AP_GPS/AP_GPS_NMEA.cpp]: @peterbarker san.. 
I will return to the original.

File: libraries/AP_Common/AP_Common.cpp (hunk 1 above)
Comment (muramura) [reply libraries/AP_Common/AP_Common.cpp]: @peterbarker san.... thanks Comments.
1. Change the name to unhex.
2. Change the method type to uint8_t.
3. Let's make the process easy. You can reduce the same process if you use lower case letters and upper case letters. Such. .

**Comment by muramura** [comment]
@peterbarker san... 
I reflected the review comments.

**Comment by peterbarker** [comment]
Needs a rebase.

**Comment by muramura** [comment]
@peterbarker san… I rebased.

**Comment by muramura** [comment]
@khancyr san.. I will change.

**Comment by tridge** [comment]
unhex isn't a great name, hex_to_char() would be better

**Comment by peterbarker** [comment]
Sorry @muramura - apparently my taste in function names isn't particularly common.

File: libraries/AP_Common/AP_Common.cpp [hunk 3]
```diff
@@ -36,3 +36,16 @@ bool is_bounded_int32(int32_t value, int32_t lower_bound, int32_t upper_bound)
...
+
+/**
+ * return the numeric value of an ascii hex character
+ *
+ * @param [in] a hexa character
+ * @return numeric value
+ * @Note Input character is 0-9, A-F, a-f
+ *  A 0x41, a 0x61, 0 0x30
+ */
+ uint8_t unhex(int8_t a)
```
Comment (tridge) [inline libraries/AP_Common/AP_Common.cpp]: not all platforms have signed characters

**Comment by tridge** [comment]
should take uint8_t not int8_t as not all platforms have signed characters

**Comment by rmackay9** [comment]
I'm going to leave the DevCallTopic on this one so that it comes up again next week.  I'm doing this because we talked about it enough that I don't want that time to be wasted.

**Comment by WickedShell** [comment]
This changes the failure semantics when an out of range input is handed to the function. Given the actual usage of this is to check for checksums we should change this to have a signature more like `bool hex_to_uint8(uint8_t a, uint8_t &res)` where we return false if it's out of range and check it. Otherwise we could pass the checksum by accident when handling corrupted data.

**Comment by rmackay9** [comment]
@muramura, I think this is waiting on you to fix this up.

**Comment by peterbarker** [comment]
```
UTC2331 - https://github.com/ArduPilot/ardupilot/pull/10941
Need the signed/unsigned char thing fixes
Need to address Michael’s concern
```

**Comment by tridge** [comment]
@muramura ping?

**Comment by muramura** [comment]
TO: Members, Contributors
Thank you for your multiple comments. I will change.

**Comment by muramura** [comment]
I changed about the following.
1. Change the method name to hex_to_uint8.
2. Change the type of input argument to uint8_t.
3. Return the result of method processing.
4. Return the conversion result as an argument.
5. Change the reference processing to the hex_to_uint8 method specification.

**Comment by khancyr** [comment]
it looks good to go ! Comments have been adressed

File: libraries/AP_GPS/AP_GPS_NMEA.cpp [hunk 4]
```diff
@@ -234,7 +221,12 @@ bool AP_GPS_NMEA::_term_complete()
 {
     // handle the last term in a message
     if (_is_checksum_term) {
-        uint8_t checksum = 16 * _from_hex(_term[0]) + _from_hex(_term[1]);
+        uint8_t nibble_high = 0;
+        uint8_t nibble_low  = 0;
+        if (hex_to_uint8(_term[0], nibble_high)  == false || hex_to_uint8(_term[1], nibble_low)  == false) {
```
Comment (rmackay9) [inline libraries/AP_GPS/AP_GPS_NMEA.cpp]: can you change this to "if (!hex_to_uint8..."

File: libraries/AP_RangeFinder/AP_RangeFinder_NMEA.cpp [hunk 5]
```diff
@@ -134,7 +134,12 @@ bool AP_RangeFinder_NMEA::decode_latest_term()
 {
     // handle the last term in a message
     if (_term_is_checksum) {
-        uint8_t checksum = 16 * char_to_hex(_term[0]) + char_to_hex(_term[1]);
+        uint8_t nibble_high = 0;
+        uint8_t nibble_low  = 0;
+        if (hex_to_uint8(_term[0], nibble_high)  == false || hex_to_uint8(_term[1], nibble_low)  == false) {
+            return false;
+        }
+        uint8_t checksum = 16 * nibble_high + nibble_low;
```
Comment (tridge) [inline libraries/AP_RangeFinder/AP_RangeFinder_NMEA.cpp]: I'd prefer (nibble_high<<4)

**Comment by rmackay9** [comment]
@muramura, can you split up the commits so that each commit only affects one library?

We discussed this on the dev call and after the above two fixes we will merge, thanks!

**Comment by khancyr** [comment]
rebase, split commit, and address comments. Ready to merge !

**Comment by OXINARF** [comment]
@khancyr If you can, please fix the commit messages.

**Comment by khancyr** [comment]
@OXINARF done !

**Comment by tridge** [comment]
@muramura has this been tested with the SITL NMEA GPS?

**Comment by khancyr** [comment]
@tridge tested on SITL with SIM_GPS_TYPE 5 and GPS_TYPE 5, and it works as master. We need to ajust the hdop in SITL to less to 1.4 otherwise the prearm check trigger as the default value is 2.0

**Comment by muramura** [comment]
@tridge san.. I have not tested.
@khancyr  san.  Thank you for the test.
